python/ray/experimental/client/__init__.py

```diff
@@ -6,55 +6,69 @@
 
 logger = logging.getLogger(__name__)
 
+# _client_api has to be external to the API stub, below.
+# Otherwise, ray.remote() that contains ray.remote()
+# contains a reference to the RayAPIStub, therefore a
+# reference to the _client_api, and then tries to pickle
+# the thing.
 _client_api: Optional[APIImpl] = None
 
 
-def check_client_api():
-    global _client_api
-    if _client_api is None:
-        logger.info(
-            "No client API initialized: probably a worker, using core ray")
-        from ray.experimental.client.core_ray_api import set_client_api_as_ray
-        set_client_api_as_ray()
+class RayAPIStub:
+    def connect(self, conn_str):
+        global _client_api
+        from ray.experimental.client.worker import Worker
+        _client_worker = Worker(conn_str)
+        _client_api = ClientAPI(_client_worker)
+
+    def disconnect(self):
+        global _client_api
+        if _client_api is not None:
+            _client_api.close()
+        _client_api = None
+
+    def __getattr__(self, key: str):
+        global _client_api
+        self.__check_client_api()
+        return _client_api.__getattribute__(key)
+
+    def __check_client_api(self):
+        global _client_api
+        if _client_api is None:
+            from ray.experimental.client.server.core_ray_api import CoreRayAPI
+            _client_api = CoreRayAPI()
+
+
+ray = RayAPIStub()
+
+# The following are methods in this module so that someone who
+# tries to `import ray_client as ray` -- as a module, instead of
+# `from ray_client import ray` -- as the API stub
+# still gets expected functionality.
```

Comment:
Can we actually remove the module-level methods for now? I'm concerned the getattr won't work in those cases and it will be quite confusing. Also, in the future when we integrate with ray proper for transparent client mode, this will have to change anyway.